libglusterfs/src/common-utils.c

```diff
@@ -642,6 +642,123 @@ gf_dnscache_entry_deinit(struct dnscache_entry *entry)
     GF_FREE(entry);
 }
 
+/**
+ * gf_dns_addr_cache_deinit -- cleanup resources used by struct dnscache
+ */
+void
+gf_dns_addr_cache_deinit(struct dnscache *cache)
+{
+    struct dnscache_addr_entry *dnsentry = NULL;
+    data_pair_t *dns_addr_pairs = NULL;
+
+    if (!cache) {
+        gf_msg_plain(GF_LOG_WARNING, "dnscache is NULL");
+        return;
+    }
+
+    dns_addr_pairs = cache->cache_dict->members_list;
+    while (dns_addr_pairs) {
+        dnsentry = (struct dnscache_addr_entry *)dns_addr_pairs->value->data;
+        freeaddrinfo(dnsentry->addr);
+        dns_addr_pairs = dns_addr_pairs->next;
+    }
+    gf_dnscache_deinit(cache);
+}
+
+/**
+ * gf_dnscache_entry_init -- Initialize a dnscache entry
+ *
+ * @return: SUCCESS: Pointer to an allocated dnscache entry struct
+ *          FAILURE: NULL
+ */
+
+struct dnscache_addr_entry *
+gf_dnscache_addr_entry_init()
+{
+    struct dnscache_addr_entry *entry = GF_CALLOC(
+        1, sizeof(*entry), gf_common_mt_dnscache_addr_entry);
+    return entry;
+}
+
+/**
+ * gf_dnscache_addr_entry_deinit -- Free memory used by a dnscache addr entry
+ *
+ * @entry: Pointer to deallocate
+ */
+
+void
+gf_dnscache_addr_entry_deinit(struct dnscache_addr_entry *entry)
+{
+    freeaddrinfo(entry->addr);
+    GF_FREE(entry);
+}
+
+/* Peform a dns lookup for identifier(IP address or hostname) */
+
+struct addrinfo *
+gf_dns_lookup_address_cached(char *identifier, struct dnscache *dnscache)
+{
+    struct addrinfo *addr = NULL;
+    dict_t *cache = NULL;
+    data_t *entrydata = NULL;
+    struct dnscache_addr_entry *dnsentry = NULL;
+    gf_boolean_t from_cache = _gf_false;
+    struct addrinfo hints;
+    int gai_err = 0;
+
+    if (!dnscache)
+        goto out;
+
+    cache = dnscache->cache_dict;
+
+    /* Quick cache lookup to see if we already hold it */
```

Comment:
Currently `dict_t` is implemented as a linked list, so lookups won't be so fast if the cache is big. For small caches, this should be ok.

Reply:
I think auth_address wont be too long so i believe we can use dict_t.

Reply:
I think we are also solving client addresses ? do we have some limit on the number of clients we may have ?

Anyway, for now this should be ok.